interp/run.go

```diff
@@ -2824,86 +2735,62 @@ func convertConstantValueTo(n *node, typ reflect.Type) {
-			if l > 8 {
-				panic(fmt.Sprintf("constant %s overflows int8", c.ExactString()))
-			}
-			v = reflect.ValueOf(int8(i))
-		case reflect.Int16:
-			if l > 16 {
-				panic(fmt.Sprintf("constant %s overflows int16", c.ExactString()))
-			}
-			v = reflect.ValueOf(int16(i))
-		case reflect.Int32:
-			if l > 32 {
-				panic(fmt.Sprintf("constant %s overflows int32", c.ExactString()))
-			}
-			v = reflect.ValueOf(int32(i))
-		case reflect.Int64:
-			if l > 64 {
-				panic(fmt.Sprintf("constant %s overflows int64", c.ExactString()))
-			}
-			v = reflect.ValueOf(i)
+		if l > bitlen[kind] {
+			panic(fmt.Sprintf("constant %s overflows int%d", c.ExactString(), bitlen[kind]))
 		}
-	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		v = reflect.ValueOf(i).Convert(typ)
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
 		i, _ := constant.Uint64Val(c)
 		l := constant.BitLen(c)
-		switch typ.Kind() {
-		case reflect.Uint:
-			if l > 64 {
-				panic(fmt.Sprintf("constant %s overflows uint", c.ExactString()))
-			}
-			v = reflect.ValueOf(uint(i))
-		case reflect.Uint8:
-			if l > 8 {
-				panic(fmt.Sprintf("constant %s overflows uint8", c.ExactString()))
-			}
-			v = reflect.ValueOf(uint8(i))
-		case reflect.Uint16:
-			if l > 16 {
-				panic(fmt.Sprintf("constant %s overflows uint16", c.ExactString()))
-			}
-			v = reflect.ValueOf(uint16(i))
-		case reflect.Uint32:
-			if l > 32 {
-				panic(fmt.Sprintf("constant %s overflows uint32", c.ExactString()))
-			}
-			v = reflect.ValueOf(uint32(i))
-		case reflect.Uint64:
-			if l > 64 {
-				panic(fmt.Sprintf("constant %s overflows uint64", c.ExactString()))
-			}
-			v = reflect.ValueOf(i)
-		case reflect.Uintptr:
-			if l > 64 {
-				panic(fmt.Sprintf("constant %s overflows uintptr", c.ExactString()))
-			}
-			v = reflect.ValueOf(i)
+		if l > bitlen[kind] {
+			panic(fmt.Sprintf("constant %s overflows uint%d", c.ExactString(), bitlen[kind]))
 		}
+		v = reflect.ValueOf(i).Convert(typ)
 	case reflect.Float32:
 		f, _ := constant.Float32Val(c)
-		v = reflect.ValueOf(f)
+		v = reflect.ValueOf(f).Convert(typ)
 	case reflect.Float64:
 		f, _ := constant.Float64Val(c)
-		v = reflect.ValueOf(f)
+		v = reflect.ValueOf(f).Convert(typ)
 	case reflect.Complex64:
 		r, _ := constant.Float32Val(constant.Real(c))
 		i, _ := constant.Float32Val(constant.Imag(c))
-		v = reflect.ValueOf(complex(r, i))
+		v = reflect.ValueOf(complex(r, i)).Convert(typ)
 	case reflect.Complex128:
 		r, _ := constant.Float64Val(constant.Real(c))
 		i, _ := constant.Float64Val(constant.Imag(c))
-		v = reflect.ValueOf(complex(r, i))
+		v = reflect.ValueOf(complex(r, i)).Convert(typ)
+	default:
+		// Type kind is from internal constant representation. Only use default types here.
+		switch c.Kind() {
+		case constant.Bool:
+			v = reflect.ValueOf(constant.BoolVal(c))
+		case constant.String:
+			v = reflect.ValueOf(constant.StringVal(c))
```

Comment:
These 2 cases are impossible at this point. We only have `constant.Value` values for numeric kinds. Should we still support it?

Reply:
I think that keeping it causes no harm, and I wonder also if we could come out with a constant boolean expression involving large numbers where this could be useful (not tested yet).

Reply:
I dont think these cases are handled in the ops, but that that another PR.